morpheus/_lib/src/io/loaders/rest.cpp

```diff
@@ -17,21 +17,295 @@
+    body                          = query.value("body", "");
+
+    nlohmann::json x_headers_json = query.value("x-headers", nlohmann::json());
+    for (auto& x_header_kv : x_headers_json.items())
+    {
+        x_headers.insert(std::make_pair((std::string)x_header_kv.key(), (std::string)x_header_kv.value()));
+    }
+
+    params = query.value("params", nlohmann::json());
+}
+
+http::verb get_http_verb(const std::string& method)
+{
+    http::verb verb;
+    if (method == "GET")
+    {
+        verb = http::verb::get;
+    }
+    else if (method == "POST")
+    {
+        verb = http::verb::post;
+    }
+    return verb;
+}
+
+http::request<http::string_body> construct_request(http::verb verb,
+                                                   const std::string& target,
+                                                   const std::string& query,
+                                                   const std::string& host,
+                                                   const std::string& method,
+                                                   const std::string& content_type,
+                                                   const std::string& body,
+                                                   const std::unordered_map<std::string, std::string>& x_headers)
+{
+    http::request<http::string_body> request{verb, target + query, 11};
+    request.set(http::field::host, host);
+    request.set(http::field::user_agent, BOOST_BEAST_VERSION_STRING);
+    if (method == "POST")
+    {
+        request.set(http::field::content_type, content_type);
+        request.body() = body;
+        request.prepare_payload();
+    }
+    for (auto& x_header : x_headers)
+    {
+        request.insert(x_header.first, x_header.second);
+    }
+    return request;
+}
+
+void try_get_data(beast::tcp_stream& stream,
+                  const net::ip::basic_resolver_results<tcp>& endpoint_results,
+                  http::request<http::string_body>& request,
+                  http::response<http::dynamic_body>& response,
+                  int max_retry,
+                  int interval_milliseconds)
+{
+    while (max_retry-- > 0)
+    {
+        stream.connect(endpoint_results);
+        http::write(stream, request);
+        beast::flat_buffer buffer;
+        http::read(stream, buffer, response);
+        beast::error_code ec;
+        stream.socket().shutdown(tcp::socket::shutdown_both, ec);
+        if (ec && ec != beast::errc::not_connected)
+        {
+            throw beast::system_error{ec};
+        }
+        int status = response.result_int();
+        // 503 Service Unavailable / 504 Gateway Timeout
+        if (status != 503 && status != 504)
+        {
+            break;
+        }
+        std::this_thread::sleep_for(std::chrono::milliseconds(interval_milliseconds));
+        interval_milliseconds *= 2;
+    }
+}
+
+void get_data_from_endpoint(http::response<http::dynamic_body>& response,
```

Comment:
It would be beneficial to split this function into smaller, more focused functions.

A helpful guideline is that when a function is performing operations that are not directly related to its stated purpose, these operations should be moved into separate functions. This enhances readability and maintainability.

In the case of the 'get_data_from_endpoint' function, it is not just retrieving data, but also parsing and formatting URLs. Therefore, it may be more effective to restructure the function as follows:

    get_data_from_endpoint
    parse_and_format_URL
    get_data

This approach will ensure that each function is focused on a single responsibility, making the code easier to understand and debug.